### PR TITLE
fix(toolkit): 'cdk deploy' support updates to Outputs

### LINKED ISSUE #778: cdk diff should indicate if only Output CF description added, then Cloudformation will not update.
CDK Version 0.9.0

Enhancement: cdk diff should indicate if only an Ouput Resource was Added, cloudformation will not update.

Steps: 
1. Add only Output Cloud formation description
2. run cdk deploy 
3. Get Stack was already up-to-date
4. run cdk diff

Expected:
[+] Added (Output will be updated on next Resource update)<Description>

Actual:
+] Added  <Desciption>

Example:
From Synth
*********************
_Outputs:
LoadBalancerArn:
Description: NLB
Value:
Ref: NLB
Export:
Name: 'NLBArn'_

From Terminal:
*****************************
user@ubuntu:~/sms100$ cdk deploy cdk-NLB
Starting deployment of stack cdk-NLB...
Stack was already up-to-date, it has ARN _arn:aws:cloudformation:us-west-1:123456:stack/cdk-NLB/01314e0-c15f-11e8-ab9b-50fae8123456_ cdk-sms104-NLB

user@ubuntu:~/sms100$ cdk diff cdk-NLB
[+] Added LoadBalancerArn: {"Description":"NLB","Value":{"Ref":"NLB"},"Export":{"Name":{"NLBArn"}}}''

### PR DESCRIPTION
**Pull request by rix0rrr**
If only stack Outputs are changed, CloudFormation generates a ChangeSet
that is executable but has 0 changes.

Before, we looked at the amount of changes to say there was nothing to
do, but now we look at the actual change set status to determine
whether it's an empty change set or not.

The effect is that we can now deploy updates even if only Outputs
changed. This becomes very important when the only thing changed
to a stack is an Output got added because a cross-stack reference
was taken by a downstream stack.

Fixes #778.

----

### Pull Request Checklist

* [x] Testing
  - Unit test added (prefer not to modify an existing test, otherwise, it's probably a breaking change)
  - __CLI change?:__ coordinate update of integration tests with team
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: packages/aws-cdk/test/util/fake-sdk.ts [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import sinon = require('sinon');
+import { SDK } from "../../lib/api/util/sdk";
+
+/**
+ * An SDK that allows replacing (some of) the clients
+ *
+ * Its the responsibility of the consumer to replace all calls that
+ * actually will be called.
+ */
+export class FakeSDK extends SDK {
```
Comment (eladb) [inline packages/aws-cdk/test/util/fake-sdk.ts]: `MockSDK` or `SDKMock`

File: packages/aws-cdk/lib/api/util/cloudformation.ts [hunk 2]
```diff
@@ -86,32 +86,53 @@ async function waitFor<T>(valueProvider: () => Promise<T | null | undefined>, ti
...
-        return description;
-      }
     }
+
     // tslint:disable-next-line:max-line-length
     throw new Error(`Failed to create ChangeSet ${changeSetName} on ${stackName}: ${description.Status || 'NO_STATUS'}, ${description.StatusReason || 'no reason provided'}`);
   });
+
+  if (!ret) {
+    throw new Error('Change set took too long to be created; aborting');
```
Comment (RomainMuller) [inline packages/aws-cdk/lib/api/util/cloudformation.ts]: Is this really a failure mode here? Like... some kind of a timeout?

File: packages/aws-cdk/lib/api/util/cloudformation.ts [hunk 3]
```diff
@@ -86,32 +86,53 @@ async function waitFor<T>(valueProvider: () => Promise<T | null | undefined>, ti
...
+
+/**
+ * Return true if the given change set has no changes
+ *
+ * This must be determined from the status, not the 'Changes' array on the
+ * object; the latter can be empty because no resources were changed, but if
+ * there are changes to Outputs, the change set can still be executed.
+ */
+export function changeSetHasNoChanges(description: CloudFormation.DescribeChangeSetOutput) {
+    return description.Status === 'FAILED'
```
Comment (RomainMuller) [inline packages/aws-cdk/lib/api/util/cloudformation.ts]: Is this 4-space indented?

File: packages/aws-cdk/lib/api/util/cloudformation.ts (hunk 2 above)
Comment (rix0rrr) [reply packages/aws-cdk/lib/api/util/cloudformation.ts]: What is the question? Can this really happen in practice, can the code path be taken, ...?

File: packages/aws-cdk/lib/api/util/cloudformation.ts (hunk 3 above)
Comment (rix0rrr) [reply packages/aws-cdk/lib/api/util/cloudformation.ts]: It might be.

**Pull request merged**